cdc/log.cc

```diff
@@ -416,6 +416,194 @@ utils::UUID generate_timeuuid(api::timestamp_type t) {
     return utils::UUID_gen::get_random_time_UUID_from_micros(t);
 }
 
+struct atomic_column_update {
+    column_id id;
+    atomic_cell cell;
+};
+
+struct nonatomic_column_deletion {
+    column_id id;
+};
+
+struct nonatomic_column_update {
+    column_id id;
+    std::vector<std::pair<bytes, atomic_cell>> cells;
+};
+
+struct static_row_update {
+    gc_clock::rep ttl;
+    std::vector<atomic_column_update> atomic_entries;
+    std::vector<nonatomic_column_deletion> nonatomic_deletions;
+    std::vector<nonatomic_column_update> nonatomic_updates;
+};
+
+struct clustered_row_insert {
+    gc_clock::rep ttl;
+    clustering_key key;
+    std::vector<atomic_column_update> atomic_entries;
+    std::vector<nonatomic_column_deletion> nonatomic_deletions;
+};
+
+struct clustered_row_update {
+    gc_clock::rep ttl;
+    clustering_key key;
+    std::vector<atomic_column_update> atomic_entries;
+    std::vector<nonatomic_column_deletion> nonatomic_deletions;
+    std::vector<nonatomic_column_update> nonatomic_updates;
+};
+
+struct clustered_row_deletion {
+    clustering_key key;
+};
+
+struct clustered_range_deletion {
+    clustering_key_prefix start;
+    bound_kind start_kind;
+    clustering_key_prefix end;
+    bound_kind end_kind;
+};
+
+struct batch {
+    std::vector<static_row_update> static_updates;
+    std::vector<clustered_row_insert> clustered_inserts;
+    std::vector<clustered_row_update> clustered_updates;
+    std::vector<clustered_row_deletion> clustered_row_deletions;
+    std::vector<clustered_range_deletion> clustered_range_deletions;
+    bool has_partition_deletion = false;
+};
+
+struct row_update {
+    std::vector<atomic_column_update> atomic_entries;
+    std::vector<nonatomic_column_deletion> nonatomic_deletions;
+    std::vector<nonatomic_column_update> nonatomic_updates;
+};
+
+using set_of_changes = std::map<api::timestamp_type, batch>;
+
+set_of_changes extract_changes(const mutation& base_mutation, const schema& base_schema) {
+    auto extract_row_updates = [&base_schema] (const row& r, column_kind ckind) {
+        std::map<std::pair<api::timestamp_type, gc_clock::rep>, row_update> ups;
+        r.for_each_cell([&] (column_id id, const atomic_cell_or_collection& cell) {
+            auto& cdef = base_schema.column_at(ckind, id);
+            if (cdef.is_atomic()) {
+                auto view = cell.as_atomic_cell(cdef);
+                auto timestamp_and_ttl = std::make_pair(
+                        view.timestamp(),
+                        view.is_live_and_has_ttl() ? view.ttl().count() : 0
+                    );
+                ups[timestamp_and_ttl].atomic_entries.push_back({id, atomic_cell(*cdef.type, view)});
```

Comment:
Is `entries` meant to contain both insertions, updates and deletions? If yes, why is this handled separately for nonatomic cells (into different fields)?

Reply:
I'd have to store an `atomic_cell_or_collection` there, and dispatch+unpack again in `transform` (i.e. call `as_atomic_cell` or `as_collection_mutation`, depending on `is_atomic()`). No need to do that, the proposed way is more convenient.

And I have to unpack nonatomics right there because the contained cells might have different timestamps and ttls.

The idea of `extract_changes` is to do all the unpacking, and later in `transform` we just consume the unpacked changes.